---
title: "Reads QC and trimming"
date: "`r format(Sys.time(), '%d %B %Y')`"
author: "Vincent Manzanilla, INTERTRYP, IRD"
output: 
  rmdformats::material:
    code_folding: show
    use_bookdown: true
    thumbnails: false
    number_sections: true
    css: "../../ETBII.css"  
header-includes:
  - \usepackage{subfig}
  - \usepackage{booktabs}
  - \usepackage{xcolor}
editor_options: 
  markdown: 
    wrap: 150
---

<!-- Custom title with logo -->
<div class="title-container">
  <img src="../../data/IRD.png" alt="IRD Logo" class="logo">
</div>

```{=html}
<style>
.textbox {
  border: 1px solid #ccc;
  padding: 10px;
  background-color: #f5f5f5;
}

.highlight {
  background-color: yellow;
  font-weight: bold;
}
</style>
```

# Design Metabarcoding

Illumina 
- read length: 300 bp

| **Category**                  | **Primer**                   | **Sequence**                                   | **Length** |
|-------------------------------|------------------------------|------------------------------------------------|------------|
| **Bacteria (142 bp)**          | E786F _16s_bac               | GATTAGATACCCTGGTAG                             | 18         |
|                               | U926R_16s_bac                | CCGTCAATTCCTTTRAGTTT                           | 20         |
| **Mammals (130 bp)**           | 16Smamífero_F                | CGGTTGGGGTGACCTCGGA                            | 19         |
|                               | 16Smamífero_R                | GCTGTTATCCCTAGGGTAACT                          | 21         |
| **Plants (504-632 bp)**        | trnL/c_plantae               | GAAATCGGTAGACGCTACG                            | 19         |
|                               | trnL/d_plantae               | GGGGATAGAGGGACTTGAAC                           | 20         |
| **Bacteria (440-460 bp)**      | Bakt_341F_16s                | CCTACGGGNGGCWGCAG                              | 17         |
| *(Sinclair et al. 2015)*       | Bakt_805R_16s                | GACTACHVGGGTATCTAATCC                          | 21         |
| **Vertebrate (215 bp)**        | L1085_12s                    | CCCAAACTGGGATTAGATACCC                         | 22         |
| *(Kitano et al. 2007)*         | H1259_12s                    | GTTTGCTGAAGATGGCGGTA                           | 20         |
| **Trypanosoma brucei (114 bp)**| 18s_Tbrucei-F                | GCGATGGATGACTTGGCTTC                           | 20         |
|                               | 18s_Tbrucei-R                | TCCCATGCGCCGTTTGCGTTC                          | 21         |
| **Trypanosoma brucei (370-684 bp)**| TRYP1_Tbrucei-S            | CGTCCCTGCCATTTGTACACAC                         | 22         |
|                               | TRYP1_Tbrucei-R              | GGAAGCCAAGTCATCCATCG                           | 20         |
| **Glossina gambiensis (246 bp)**| ITS-1-GlossPalGam-F          | TGGACTTCGGATTAAGTACAACA                        | 23         |
|                               | ITS-1-GlossPalGam-R          | TCATTATGCGCTATTAAGGTAAGC                       | 24         |




# Define list of samples


```{bash, eval=F}
cd /home/vincent/Documents/metagenomics_conakry/data/fastq

# Store filenames in an array
filenames=(*.fastq.gz)

SAMPLES=()
for f in "${filenames[@]}"; do
  SAMPLES+=("$(echo "$f" | sed 's/_R[0-9]\+_001\.fastq.gz//')")
done

# Ensure unique samples
SAMPLES=($(echo "${SAMPLES[@]}" | tr " " "\n" | sort -u | tr "\n" " "))

# Print the unique samples
echo "${SAMPLES[@]}"
```

# Run Cutadapt


```{bash, eval=F}
# Create output folder
mkdir -p TRIMMED

# Run Cutadapt
for SAMPLE in "${SAMPLES[@]}"; do 
  cutadapt "$SAMPLE"_R1_001.fastq.gz \
           "$SAMPLE"_R2_001.fastq.gz \
           -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA \
           -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT \
           -o TRIMMED/"$SAMPLE".R1.fastq.gz \
           -p TRIMMED/"$SAMPLE".R2.fastq.gz \
           --minimum-length 280 \
           --quality-cutoff 20 \
           --cores 16 > TRIMMED/"$SAMPLE".cutadapt.log.txt
done
```

# run fastQC and multiQC

```{bash, eval=F}
mkdir -p QC
# Run FastQC on the trimmed reads
for SAMPLE in "${SAMPLES[@]}"; do
  fastqc TRIMMED/"$SAMPLE".R1.fastq.gz TRIMMED/"$SAMPLE".R2.fastq.gz -o QC
done

# Run MultiQC to aggregate FastQC results
multiqc QC -o QC/multiqc_report
```

# Cutadap report / trimming report

**length:** The length of the sequence.
**count:** The number of reads with this length.
**expect:** The expected error rate for detecting the adapter in sequences of this length.

```{r}
# Create the data frame
df <- data.frame(
  length = c(3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 199, 201, 203, 205, 208, 224, 229, 232, 242, 247, 250, 294, 298, 301),
  count = c(134, 10, 6, 4, 2, 5, 19, 15, 21, 11, 4, 22, 6, 20, 7, 31, 15, 8, 15, 11, 14, 22, 13, 2, 4, 12, 11, 12, 14, 11, 10, 18, 17, 16, 19, 8, 18, 29, 35, 10, 37, 8, 9, 8, 8, 7, 21, 14, 16, 16, 16, 42, 41, 25, 37, 29, 27, 35, 47, 30, 26, 42, 47, 54, 31, 60, 95, 75, 107, 160, 254, 602, 1332, 17180, 13707, 5384, 3483, 6469, 4027, 4085, 2826, 2181, 1538, 1639, 1050, 642, 436, 240, 214, 161, 110, 77, 47, 44, 38, 33, 25, 20, 42, 28, 36, 42, 35, 36, 35, 16, 15, 19, 16, 26, 16, 15, 19, 13, 13, 15, 10, 10, 17, 15, 19, 17, 26, 36, 38, 29, 38, 81, 123, 117, 286, 1100, 155, 67, 59, 84, 136, 72, 25, 20, 13, 16, 15, 11, 4, 7, 10, 11, 20, 18, 32, 31, 40, 51, 66, 159, 602, 38, 21, 8, 5, 4, 8, 6, 9, 7, 10, 17, 14, 9, 10, 4, 6, 17, 38, 13, 14, 14, 17, 19, 24, 32, 22, 83, 361, 41, 24, 12, 7, 3, 2, 2, 1, 3, 4, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1),
  expect = c(2080.3, 520.1, 130, 32.5, 8.1, 2, 0.5, 0.1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
)

# View the first few rows of the data
head(df)

```
```{r}
# Create a barplot of the counts vs. length
library(ggplot2)

ggplot(df, aes(x = factor(length), y = count)) +
  geom_bar(stat = "identity", fill = "skyblue") +
  labs(title = "Count by Length", x = "Length", y = "Count") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 8)) +  # Set smaller text size
  scale_x_discrete(breaks = seq(0, 301, by = 5)) 

```
**Q: What do you observe?**
